designs/node-repair.md

```diff
@@ -0,0 +1,92 @@
+
+* Mega Issue: https://github.com/kubernetes-sigs/karpenter/issues/750
+    * Related (Unreachable): https://github.com/aws/karpenter-provider-aws/issues/2570
+    * Related (Remove by taints): https://github.com/aws/karpenter-provider-aws/issues/2544
+    * Related (Known resource are not registered) Fixed by v0.28.0: https://github.com/aws/karpenter-provider-aws/issues/3794
+    * Related (Stuck on NotReady): https://github.com/aws/karpenter-provider-aws/issues/2439
+
+#### Out of scope
+
+The scope of the solution suggested bellow will be an opinionated method that will not give customer configurability. The team does not have enough data to determine the right level of configuration surface that users will utilize. The implementation will not consider budgets, API surface for repairing unhealthy nodes, or option for nodes to be rebooted instead of replaced. The solution will also not give any Karpenter based approach to enable AZ resiliency. **The feature will be gated under an alpha feature flag to allow for additional feedback from customers, and make subsequent changes in the future.**
+
+### Option  1 (recommended) : Unhealthy condition set cloud provider interface  
+
+```
+type UnhealthyConditions struct {
+    // Type of unhealthy state that is found on the node
+    Type string 
+    // Status condition of when a node is unhealthy
+    status string
+    // RepairAfter is the duration the controller will wait
+    // before attempting to terminate nodes that are underutilized.
+    RepairAfter time.Duration
+}
+
+type CloudProvider interface {
+  ...
+    // UnhealthyConditions is for CloudProviders to define a set Unhealthy condition for Karpenter 
+    // to monitor on the node. Customer will need 
+    UnhealthyConditions() []v1.UnhealthyConditions
+  ...
+}
+```
+
+The `UnhealthyConditions` will be a set of condition that the Karpenter controller will watch. The cloud provider can define compatibility with any node diagnostic agent, and track a list of node unhealthy condition types and a duration period to wait until a unhealthy state is considered a terminal: 
+
+1. A diagnostic agent will add a status condition on to a node 
+2. Karpenter will reconcile on nodes and match unhealthy node condition to cloud provider defined unhealthy condition 
```

Comment:
Does this imply that the cloud provider needs to be aware of the node being unhealthy?
We observed cases where the node in our EKS cluster was in an unreachable state (kubelet died) but the from the AWS console point of view it was fine (health-checks passing).

Will this implementation consider the case where the node is unhealthy from kubernetes point of view but healthy from the cloud provider side?

Reply:
Yes, all the solution here focus on Kubernetes status condition. We plan on acting when a node is unhealthy from a Kubernetes point of view or an agent has marked as such. Cloud providers need to have a supported node diagnostic agent like [node-problem-detector](https://github.com/kubernetes/node-problem-detector). A cloud provider will specify the node condition that Karpenter will watch for and then act against.